xfuser/model_executor/models/transformers/hunyuan_transformer_2d.py

```diff
@@ -0,0 +1,237 @@
+from typing import Optional, Dict, Any
+import torch
+import torch.distributed
+import torch.nn as nn
+
+from diffusers import HunyuanDiT2DModel
+from diffusers.models.embeddings import PatchEmbed
+from diffusers.models.attention_processor import Attention
+from diffusers.models.transformers.transformer_2d import Transformer2DModelOutput
+from diffusers.utils import is_torch_version
+
+from xfuser.logger import init_logger
+from xfuser.distributed.runtime_state import get_runtime_state
+from xfuser.model_executor.base_wrapper import xFuserBaseWrapper
+from xfuser.distributed import (
+    get_pipeline_parallel_rank,
+    get_pipeline_parallel_world_size,
+    is_pipeline_first_stage,
+    is_pipeline_last_stage,
+)
+from .register import xFuserTransformerWrappersRegister
+from .base_transformer import xFuserTransformerBaseWrapper
+
+logger = init_logger(__name__)
+
+
+@xFuserTransformerWrappersRegister.register(HunyuanDiT2DModel)
+class xFuserHunyuanDiT2DWrapper(xFuserTransformerBaseWrapper):
+    def __init__(
+        self,
+        transformer: HunyuanDiT2DModel,
+    ):
+        super().__init__(
+            transformer=transformer,
+            submodule_classes_to_wrap=[nn.Conv2d, PatchEmbed, Attention],
```

Comment:
attn2 in HunyuanDiTBlock is a cross-attention. It uses a pre-computed tensor, a.k.a encoder_hidden_states=encoder_hidden_states as kv? do we need to warp attn2 here?


https://github.com/huggingface/diffusers/blob/98930ee131b996c65cbbf48d8af363a98b21492c/src/diffusers/models/transformers/hunyuan_transformer_2d.py#L192C13-L192C57

Reply:
Yes, because even in cross-attention we need to apply rotary embedding to query, the query is sliced to a patch, so the rotary embedding need to be sliced as well. We need to slice the rotary embedding in the attention processer, because self attention requires whole rotary embedding for key, so the rotary embedding input to HunyuanDiTBlock need to be whole.